core/src/peer_record.rs

```diff
@@ -0,0 +1,254 @@
+use crate::identity::Keypair;
+use crate::signed_envelope::SignedEnvelope;
+use crate::{peer_record_proto, signed_envelope, Multiaddr, PeerId};
+use std::convert::TryInto;
+use std::fmt;
+
+const PAYLOAD_TYPE: &str = "/libp2p/routing-state-record";
+const DOMAIN_SEP: &str = "libp2p-routing-state";
+
+// TODO: docs
+#[derive(Debug, PartialEq, Clone)]
+pub struct PeerRecord {
+    pub peer_id: PeerId,
+    pub seq: u64,
+    pub addresses: Vec<Multiaddr>,
+}
+
+impl PeerRecord {
+    // TODO: docs
+    pub fn into_protobuf_encoding(self) -> Vec<u8> {
+        use prost::Message;
+
+        let record = peer_record_proto::PeerRecord {
+            peer_id: self.peer_id.to_bytes(),
+            seq: self.seq,
+            addresses: self
+                .addresses
+                .into_iter()
+                .map(|m| peer_record_proto::peer_record::AddressInfo {
+                    multiaddr: m.to_vec(),
+                })
+                .collect(),
+        };
+
+        let mut buf = Vec::with_capacity(record.encoded_len());
+        record
+            .encode(&mut buf)
+            .expect("Vec<u8> provides capacity as needed");
+        buf
+    }
+
+    pub fn from_protobuf_encoding(bytes: &[u8]) -> Result<Self, DecodingError> {
+        use prost::Message;
+
+        let record = peer_record_proto::PeerRecord::decode(bytes)?;
+
+        Ok(Self {
+            peer_id: PeerId::from_bytes(&record.peer_id)?,
+            seq: record.seq,
+            addresses: record
+                .addresses
+                .into_iter()
+                .map(|a| a.multiaddr.try_into())
+                .collect::<Result<Vec<_>, _>>()?,
+        })
+    }
+
+    // TODO: docs
+    pub fn warp_in_envelope(self, key: Keypair) -> SignedEnvelope {
+        if key.public().into_peer_id() != self.peer_id {
+            panic!("bad key")
+        }
+
+        let payload = self.into_protobuf_encoding();
+
+        SignedEnvelope::new(
+            key,
+            String::from(DOMAIN_SEP),
+            PAYLOAD_TYPE.as_bytes().to_vec(),
+            payload,
+        )
+        .unwrap() // TODO: Error handling
+    }
+
+    pub fn authenticate(self, key: Keypair) -> AuthenticatedPeerRecord {
```

Comment:
Yes, agreed that the indirections are not good. I wasn't quite sure what the most natural and expressive way of constructing one of these would be.

Will see if we can make this work using From/TryFrom for the internals. I'd like to keep domain-specific functions for the actual public API because I consider them more expressive.

Reply:
I removed `PeerRecord::authenticate`.

I think `PeerRecord::wrap_in_envelope` should stay as the information about the domain-separator should be tied to the `PeerRecord` because a `SignedEnvelope` is more abstract than a `PeerRecord`.

Reply:
What might be debatable is whether or not we even need the concept of a normal `PeerRecord`. There is probably no point in having an "unauthenticated" peer record.